libcodechecker/report.py

```diff
@@ -0,0 +1,444 @@
+        or not supported by the analyzer.
+        """
+        # TODO: REVIEW IS NEEDED WHAT TO INCLUDE IN THE HASH.
+        # NOTE: DO NOT USE UNTIL COMMENT IS REMOVED!
+
+        source_file = self.main_section.start_range.begin.file_path
+        source_line = self.main_section.start_range.begin.line
+
+        from_col = self.main_section.start_range.begin.col
+        until_col = self.main_section.end_range.begin.col
+
+        from_col = self.main_section.location.col
+        until_col = self.main_section.location.col
+
+        line_content = linecache.getline(source_file, source_line)
+
+        if line_content == '' and not os.path.isfile(source_file):
+            LOG.error("Failed to generate report hash.")
+            LOG.error('%s does not exists!' % source_file)
+
+        file_name = os.path.basename(source_file)
+        msg = self.main_section.msg
+
+        # Add the main diag section for the hash content.
+        hash_content = [file_name,
+                        self.checker_name,
+                        msg,
+                        line_content,
+                        str(from_col),
+                        str(until_col)]
+
+        diag_sections = self.__bug_path.get_diag_sections()
+
+        # Add only the control diag section column numbers to the hash content.
+        ctrl_sections = [x for x in diag_sections if x.kind == "control"]
+
+        for diag_section in ctrl_sections:
+            hash_content.append(str(diag_section.start_range.begin.col))
+            hash_content.append(str(diag_section.start_range.end.col))
+            hash_content.append(str(diag_section.end_range.begin.col))
+            hash_content.append(str(diag_section.end_range.end.col))
+
+        string_to_hash = '|||'.join(hash_content)
+
+        bug_hash = hashlib.md5(string_to_hash.encode()).hexdigest()
+        return bug_hash
+
+
+class GenericEquality(object):
+
+    def __eq__(self, other):
+        return (isinstance(other, self.__class__) and
+                self.__dict__ == other.__dict__)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+
+class Position(GenericEquality):
+    """
+    Represent a postion in a source file.
+    """
+
+    def __init__(self, line, col, filepath):
+        self.line = line
+        self.col = col
+        self.file_path = filepath
+
+    def __str__(self):
+        msg = "@"+str(self.line)+":"+str(self.col)+"::"+str(self.file_path)
+        return msg
+
+
+class Range(object):
+    """
+    Two positions build a range.
+    """
+    def __init__(self, begin, end=None):
+        self.__begin = begin
+        if end:
+            self.__end = end
+        else:
+            self.__end = begin
+
+    def __str__(self):
+        msg = "["+str(self.__begin) + "][" + str(self.__end)+"]"
+        return msg
+
+    @property
+    def begin(self):
+        return self.__begin
+
+    @property
+    def end(self):
+        return self.__end
+
+
+class BugPath(object):
```

Comment:
I wanted to create some separation from the path independent diagnostic sections ...
I will refactor this, so the path can be built only once and no element can be added later which might change the path hash and move the path hash generation here.